ios/NotificationServiceExtension/NotificationService.swift

```diff
@@ -0,0 +1,265 @@
+// Copyright © 2022-2024 IT ENGINEERING MANAGEMENT INC,
+//                       <https://github.com/team113>
+//
+// This program is free software: you can redistribute it and/or modify it under
+// the terms of the GNU Affero General Public License v3.0 as published by the
+// Free Software Foundation, either version 3 of the License, or (at your
+// option) any later version.
+//
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License v3.0 for
+// more details.
+//
+// You should have received a copy of the GNU Affero General Public License v3.0
+// along with this program. If not, see
+// <https://www.gnu.org/licenses/agpl-3.0.html>.
+
+import FirebaseMessaging
+import SQLite
+import UserNotifications
+
+class NotificationService: UNNotificationServiceExtension {
+  var contentHandler: ((UNNotificationContent) -> Void)?
+  var bestAttemptContent: UNMutableNotificationContent?
+
+  override func didReceive(
+    _ request: UNNotificationRequest,
+    withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void
+  ) {
+
+    self.contentHandler = contentHandler
+    bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
+
+    if let bestAttemptContent = bestAttemptContent {
+      let userInfo = request.content.userInfo
+
+      Task {
+        if (userInfo["chatId"] != nil) {
+          await acknowledgeDelivery(chatId: userInfo["chatId"] as! String)
+        }
+      }
```

Comment:
Ok, the only thing missing to implement is to understand why this code is being invoked only for notifications without an image. Perhaps it's because Firebase's `Messaging.serviceExtension` executes `contentHandler` right away, which finishes the execution of this extension (since no image is attached, Firebase doesn't load anything and invokes the default handler)?

Other than that, everything seems to be working fine. When a notification with image is sent over, client takes the credentials from SQLite's database. If those are expired, then it fetches the new ones from GraphQL endpoint and stores them back. And then it sends the `Chat.items(first: 1)` query to mark the messages as delivered.

A thing to note: as iOS's application and its NotificationServiceExtension are different executables, they don't share neither documents nor any values. So I've used the App Groups feature to store the SQLite database there and then access it without any problems from the extension. The same goes for the GraphQL endpoint URL.